### ISSUE TITLE
.jshintrc's globals and predef options aren't powerful enough

### ISSUE THREAD

**Issue by Mingling94** [jshint/jshint#2482]
I feel like one should be able to use jshint without adding jshint-script to the code e.g. `// jshint undef:true`

This way, jshint, as another tool in the bag of tools a developer needs to learn on each project, doesn't present itself at the top of each code-file and impact learning.

Copied and pasted from an issue on a private repo I'm working on:
Linting produces the following currently:
```
app/js/UI.js: line 7, col 5, 'fs' is defined but never used.
app/js/UI.js: line 8, col 5, 'path' is defined but never used.

app/js/pluginManager.js: line 5, col 1, 'UI' is not defined.
app/js/pluginManager.js: line 8, col 21, 'path' is not defined.
app/js/pluginManager.js: line 15, col 9, 'fs' is not defined.
app/js/pluginManager.js: line 37, col 25, 'path' is not defined.
app/js/pluginManager.js: line 50, col 27, 'path' is not defined.
app/js/pluginManager.js: line 66, col 31, 'path' is not defined.

8 errors
```
and when these variables are added to .jshintrc options as 'global' or 'predef', they get flagged as being redefined:
```
app/js/UI.js: line 7, col 5, Redefinition of 'fs'.
app/js/UI.js: line 8, col 5, Redefinition of 'path'.
app/js/UI.js: line 11, col 5, Redefinition of 'UI'.
app/js/UI.js: line 7, col 5, 'fs' is defined but never used.
app/js/UI.js: line 8, col 5, 'path' is defined but never used.

5 errors
```

Having to edit the code itself is not acceptable because .jshintrc is a development tool. Having .jshint-script be in the code when the distributables themselves don't come with jshint is harmless, but illogical and obfuscates the code from the POV of a new dev coming on-board

**Comment by lukeapage** [comment]
```
/* jshint unused: strict */
/* globals a: true, b: true, c: true */
/* predef d: true, e: true, f: true */

a = 1;
var b = 1;
if (c) {
}
d = 1;
var e = 1;
if (f) {
}
```

There is no way if you are doing `var a =` in the global scope to make a global a - this looks like a problem to me.

If you do `window.a =` or `global.a =` or you do `a =` then you can use globals `a: true` in your `.jshintrc`

The real problem is unused. If you var a variable in the global scope and are not using modules, then you are defining a global, so it isn't really unused.

**Comment by jugglinmike** [comment]
I'm having a little trouble understanding your problem. It sounds like you expect the global variables defined in one file to be accepted in another. If this is the case, have you tried the `--prereq` [CLI flag](http://jshint.com/docs/cli/)?

> #### `--prereq`
>
> Allows you to specify prerequisite files i.e. files which include definitions of global variables used throughout your project.

**Comment by Mingling94** [comment]
Hmm, I'm having trouble using --prereq, please advise?

From within my package.json:
`	"lint": "jshint app/. --exclude app/libs/ --prereq app/js/UI.js",`
I get 
```
ming@ming-Lenovo-Yoga-2-Pro:~/Github/test-ui$ npm run lint

> Test-UI@0.4.0 lint /home/ming/Github/test-ui
> jshint app/. --prereq app/js/UI.js --exclude app/libs/

ERROR: Unknown option --prereq
npm ERR! weird error 1
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! not ok code 0
```

**Comment by Mingling94** [comment]
Additionally, why have this as only a CLI option? Can one not have it in the .jshintrc as well?

It just seems a bit messy to have to edit my package.json as well as my .jshintrc for this linting tool

**Comment by jugglinmike** [comment]
Oh, then good news! `prereq` is already implemented as a configuration; we have [an open pull request to implement the CLI option](https://github.com/jshint/jshint/pull/2474) (I got those backwards). Sorry for the confusion.

Does this solve your problem?

**Comment by Mingling94** [comment]
Could you link me to where `prereq` is defined in the documentation? Are you sure you're not talking about predef?

## Context
My usecase is that I have a A.js file that declares variables, `fs = require('fs');` and `path = require('path');` for example, shared among many plugins. One such plugin, B.js, uses those variables like:
```js
A._B = (function() {
    function init() {
        fs.doStuff();
    }
}
```

## The problem
Currently JSHint either flags `A`, `fs` and `path` as not defined in B.js like so:
```
app/js/pluginManager.js: line 5, col 1, 'UI' is not defined.
app/js/pluginManager.js: line 8, col 21, 'path' is not defined.
app/js/pluginManager.js: line 15, col 9, 'fs' is not defined.
```
or as a `Redfinition` in A.js like so:
```
app/js/UI.js: line 7, col 5, Redefinition of 'fs'.
app/js/UI.js: line 8, col 5, Redefinition of 'path'.
app/js/UI.js: line 11, col 5, Redefinition of 'UI'.
```
if I define them in my `.jshintrc` under `globals` or `predef`

**Comment by jugglinmike** [comment]
Surprising status on this one: the `prereq` *option* is implemented but not documented, while the `--prereq` *flag* is documented but not implemented. [We do maintain tests for the option](https://github.com/jshint/jshint/blob/2444a0463e1a99d46e4afa50ed934c317265529d/tests/cli.js#L182-L202), though, so it's not completely informal.

Would you mind giving this a try in your JSHint configuration:

```json
{
    "prereq": ["apps/js/UI.js"]
}
```

I've opened https://github.com/jshint/jshint.github.io/issues/26 to address the documentation problem.

**Comment by Mingling94** [comment]
I've since added some code and files to my repo, but that change inflicted weird responses. It changed the lint output from:
```bash
ming@ming-Lenovo-Yoga-2-Pro:~/Github/test-ui$ npm run lint

> Test-UI@0.4.0 lint /home/ming/Github/test-ui
> jshint app/. --exclude app/libs/ --verbose

app/js/UI.js: line 7, col 5, 'fs' is defined but never used. (W098)
app/js/UI.js: line 8, col 5, 'path' is defined but never used. (W098)

app/js/plugins.js: line 5, col 1, 'UI' is not defined. (W117)
app/js/plugins.js: line 7, col 22, 'path' is not defined. (W117)
app/js/plugins.js: line 21, col 9, 'fs' is not defined. (W117)
app/js/plugins.js: line 33, col 28, 'path' is not defined. (W117)
app/js/plugins.js: line 34, col 21, 'UI' is not defined. (W117)
app/js/plugins.js: line 76, col 23, 'path' is not defined. (W117)
app/js/plugins.js: line 92, col 13, 'UI' is not defined. (W117)

app/plugins/Library/main.js: line 5, col 1, 'UI' is not defined. (W117)
app/plugins/Library/main.js: line 8, col 15, 'path' is not defined. (W117)
app/plugins/Library/main.js: line 9, col 16, 'path' is not defined. (W117)

app/plugins/Overview/main.js: line 5, col 1, 'UI' is not defined. (W117)
app/plugins/Overview/main.js: line 8, col 15, 'path' is not defined. (W117)
app/plugins/Overview/main.js: line 9, col 16, 'path' is not defined. (W117)

app/plugins/Wallet/main.js: line 5, col 1, 'UI' is not defined. (W117)
app/plugins/Wallet/main.js: line 8, col 15, 'path' is not defined. (W117)
app/plugins/Wallet/main.js: line 9, col 16, 'path' is not defined. (W117)

18 errors
```
to this:
```bash
ming@ming-Lenovo-Yoga-2-Pro:~/Github/test-ui$ npm run lint

> Test-UI@0.4.0 lint /home/ming/Github/test-ui
> jshint app/. --exclude app/libs/ --verbose

app/js/UI.js: line 39, col 1, Expected an assignment or function call and instead saw an expression. (W030)
app/js/UI.js: line 40, col 5, 'fs' is defined but never used. (W098)
app/js/UI.js: line 41, col 5, 'path' is defined but never used. (W098)

app/js/plugins.js: line 35, col 1, Expected an assignment or function call and instead saw an expression. (W030)

app/main.js: line 36, col 1, Expected an assignment or function call and instead saw an expression. (W030)
app/main.js: line 7, col 5, 'fs' is defined but never used. (W098)
app/main.js: line 8, col 5, 'path' is defined but never used. (W098)

app/plugins/Library/main.js: line 35, col 1, Expected an assignment or function call and instead saw an expression. (W030)
app/plugins/Library/main.js: line 7, col 5, 'fs' is defined but never used. (W098)

app/plugins/Overview/main.js: line 35, col 1, Expected an assignment or function call and instead saw an expression. (W030)
app/plugins/Overview/main.js: line 7, col 5, 'fs' is defined but never used. (W098)

app/plugins/Wallet/main.js: line 35, col 1, Expected an assignment or function call and instead saw an expression. (W030)
app/plugins/Wallet/main.js: line 7, col 5, 'fs' is defined but never used. (W098)

13 errors
```

I see it did reduce the errors, but a lot of those flagged line numbers don't even exist. additionally, anything defined for the plugin's main.js that isn't used, `fs`, is flagged as unused.

**Comment by Mingling94** [comment]
Essentially, the prereq option is horrendously buggy O_o

**Comment by lukeapage** [comment]
> I'm having a little trouble understanding your problem. It sounds like you expect the global variables defined in one file to be accepted in another.

It might be useful to have a `globals foo: ignore` option where if set you can use it, you can declare it and if you don't use it, you don't get a warning.

But the usefulness of such an option in the modern age is probably limited.. if you have multiple globals and reading them in one, writing in another then our current solution is either inline jshint comments to tell jshint what you want doing with each global in each file or refactoring the code to the more common module pattern where you don't use globals.

**Comment by Mingling94** [comment]
I actually did just that. I was having some variables `A.js` own others `B.js` being `A.B` and extend A's included libraries `fs = require('fs'); path = require('path');` so that B would share the libraries declared in A. I modularized things and let B require its own libraries. Then, A would include B to use it.

Not sure if this makes sense, but essentially I was previously not doing things the node way, now I am and the jshint errors are all dealt with.
